### ISSUE TITLE
Breaking change in event handler interface caused by #207

### ISSUE THREAD

**Issue by rmoreas** [irods/irods_capability_automated_ingest#219]
Adding the `scanner` and `op` as positional arguments in call to `event_handler.call()` in [`sync_irods.py` at line 535](https://github.com/irods/irods_capability_automated_ingest/blob/94b0efb478d922b5e1e65144d269904f795834a7/irods_capability_automated_ingest/sync_irods.py#L535) (see fragment below), breaks the interface of the event handler methods. These arguments should be passed as named/keyword arguments instead.

```python
        # ...
        elif content:
            if put:
                sync = op in [Operation.PUT_SYNC, Operation.PUT_APPEND]
                if sync:
                    event_handler.call(
                        "on_data_obj_modify",
                        logger,
                        sync_file,
                        logger,
                        session,
                        meta,
                        scanner,
                        op,
                        **options
                    )
```

The signature of event handler PEP methods are (e.g. `post_data_obj_modify`):

```python
def post_data_obj_modify(cls, hdlr_mod, logger, session, meta, **options)
```

Because of the above change in #207, the signature of the event handler methods need to change to:

```python
def post_data_obj_modify(cls, hdlr_mod, logger, session, meta, *args, **options)
```

In my opinion, adding `*args` makes the interface unclear. I think that all extra arguments should be passed as named arguments.

**Comment by alanking** [comment]
Good catch. We will think about how best to handle this. Seeing as the ingest tool is pre-1.0 and the interfaces are still (yes, still) settling, we didn't give much consideration for existing deployments. We will write back with our plan after UGM (and possibly after a discussion with the kind folks of KU Leuven!)

**Comment by rmoreas** [comment]
Not directly related directly related, but would be nice that `content` flag passed to `sync_data_from_file` is also passed to the `pre_data_obj_modify` and `post_data_obj_modify` handler methods to be able to differentiate between a data content change or metadata-only change.

**Comment by alanking** [comment]
To restore the interface to match the released one, we decided that we can put the new arguments in the `meta` dict. In this way, the interface only changes in the sense that some new keys appear in the dict. Besides the explicit arguments that currently exist, all the other information passed around internally and through the event handler interface is stored in `meta`, so this seemed the least disruptive.

**Comment by alanking** [comment]
@rmoreas - We've merged the changes to restore the event handler interface. I'm going to close this as resolved, but please feel free to open another issue if you have another approach you'd prefer. I think this is the one we're going to go with for 0.5.0. You're also welcome to comment further here, as well.

**Comment by rmoreas** [comment]
This `meta` parameter is the dictionary that is passed along to the Celery tasks, so I'm not sure if it's a good idea mix-in other key/values into this dictionary.

In my opinion, I think that it would be better to use keyword arguments and use the special `**options` parameter to capture the extra informal keyword parameters for the handler method in the dictionary in `options`.

**Comment by rmoreas** [comment]
Actually, I'm not a fan of either `meta` or  `**options` because using dictionaries to pass arguments are an anti-pattern.

As far as I understand the `meta` dictionary is used as argument for the celery tasks and it is also supposed to be json serializable. Therefor I think it makes the code even more confusing to reuse this `meta` dictionary to pass extra parameters to underlying calls made by the celery tasks. Also the `scanner` is instantiated by each celery task based on the settings in `meta` and the `op` is determined by calling the `operation` method on the handler.

Maybe it's possible to rename `**options` to `**kwargs`? Then it doesn't sound as optional 😉. But I agree, it would be better to define formal named parameters.

**Comment by rmoreas** [comment]
I just find out that `scanner` and `op` are actually optional because these are only required when calling the `on_data_obj_modify` handler method for a `Operation.PUT_SYNC` or `Operation.PUT_APPEND` operation. In other cases, `scanner` and `op` are not passed.

**Comment by alanking** [comment]
Oh, interesting. This is indicative of the need for a refactor, but that's definitely a vote for using `**options`. It's easy enough to switch them, so I can do that.

**Comment by alanking** [comment]
Okay, so after some investigation and discussion, I think we've come around to where @rmoreas was in April. Thanks for your patience. :)

Adding `*args` does indeed resolve the issue, and we do not need to put `scanner` or `op` into `meta`. That said, we agree that `*args` makes the interface less clear, but implementers of the event handler methods should avoid modifying these arguments at the risk of causing the task to fail, so in that way it is advantageous to hide them. Additionally, this interface will continue to change over time as we improve the internals of this tool, so it's possible that these positional arguments will go away again in the future. The `*args` parameter will allow us to make such changes without impacting the event handler methods again in the future.

For 0.5.0, we will need to update all of our example event handler methods and documentation and training to include the new `*args` parameters. I also discovered that this issue got past our testing harness because the existing tests do not cover the use of the event handler methods with the `PUT_SYNC` or `PUT_APPEND` operations. So we will need to add tests around this as well.

Thoughts? Questions? Complaints?

Bonus note:

While investigating this, we discovered something about `**options` which offers more clarity as to its (possible) intended purpose. `**options` is passed directly to the PRC calls which are then used primarily as the "condInput" to the underlying API calls being made to the iRODS server. The only things I see presently getting inserted into `**options` in the ingest tool are iRODS keywords ("destRescName", `DATA_SIZE_KW`, etc.). I think the intent of these keyword args is to be passed to the iRODS server via the PRC calls. This caused us to steer away from adding `scanner` and `op` to `**options` as well, because these would be uselessly passed along to the iRODS server.
